src-tauri/src/telemetry_manager.rs

```diff
@@ -0,0 +1,321 @@
+use jsonwebtoken::errors::Error as JwtError;
+use jsonwebtoken::{decode, Algorithm, DecodingKey, TokenData, Validation};
+use log::{debug, error, info, warn};
+use serde::{Deserialize, Serialize};
+use std::{sync::Arc, thread::sleep, time::Duration};
+use tari_common::configuration::Network;
+use tari_core::transactions::tari_amount::MicroMinotari;
+use tokio::sync::RwLock;
+use tokio_util::sync::CancellationToken;
+
+use crate::{
+    app_config::{AppConfig, MiningMode},
+    cpu_miner::CpuMiner,
+    gpu_miner::GpuMiner,
+    hardware_monitor::HardwareMonitor,
+    node_manager::NodeManager,
+    UniverseAppState, LOG_TARGET,
+};
+
+#[derive(Debug, Deserialize, Serialize)]
+struct AirdropAccessToken {
+    exp: u64,
+    iat: i32,
+    id: String,
+    provider: String,
+    role: String,
+    scope: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum TelemetryResource {
+    Cpu,
+    Gpu,
+    #[serde(rename(serialize = "cpu-gpu"))]
+    CpuGpu,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum TelemetryNetwork {
+    #[serde(rename(serialize = "mainnet"))]
+    Main,
+    #[serde(rename(serialize = "stagenet"))]
+    Stage,
+    #[serde(rename(serialize = "nextnet"))]
+    Next,
+    #[serde(rename(serialize = "localnet"))]
+    Local,
+    #[serde(rename(serialize = "igor"))]
+    Igor,
+    #[serde(rename(serialize = "esmeralda"))]
+    Esmeralda,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TelemetryManagerError {
+    #[error("telemetry data could not be sent because of an error: {0}")]
+    ReqwestError(#[from] reqwest::Error),
+
+    #[error("Other error: {0}")]
+    Other(#[from] anyhow::Error),
+
+    #[error("Jwt decodeing error: {0}")]
+    JwtDecodeError(#[from] JwtError),
+}
+
+impl From<Network> for TelemetryNetwork {
+    fn from(value: Network) -> Self {
+        match value {
+            Network::MainNet => TelemetryNetwork::Main,
+            Network::StageNet => TelemetryNetwork::Stage,
+            Network::NextNet => TelemetryNetwork::Next,
+            Network::LocalNet => TelemetryNetwork::Local,
+            Network::Igor => TelemetryNetwork::Igor,
+            Network::Esmeralda => TelemetryNetwork::Esmeralda,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum TelemetryMiningMode {
+    Eco,
+    Ludicrous,
+}
+
+impl From<MiningMode> for TelemetryMiningMode {
+    fn from(value: MiningMode) -> Self {
+        match value {
+            MiningMode::Eco => TelemetryMiningMode::Eco,
+            MiningMode::Ludicrous => TelemetryMiningMode::Ludicrous,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TelemetryData {
+    pub app_id: String,
+    pub block_height: u64,
+    pub is_mining_active: bool,
+    pub network: Option<TelemetryNetwork>,
+    pub resource_used: TelemetryResource,
+    pub cpu_hash_rate: Option<f64>,
```

Comment:
FYI: We're planning to replace these floats with integers.

Reply:
It will probably not collide with the db, so you can. Please let me know when that happens.